# Bioinformatics

Code inspired by [Finding Hidden Messages in DNA (Bioinformatics I)](https://class.coursera.org/hiddenmessages-003) and 
from [Rosalind](http://rosalind.info). 
NB: functions generally use zero based indexing; Rosalind uses 1-based.

## Textbook track

| # | Location | Description |
| ---- | ----------- | ------------------------------------------------|
|1|-| Where in the Genome does DNA replication begin?|
|BA1A|rosalind.py|[Compute the Number of Times a Pattern Appears in a Text](http://rosalind.info/problems/ba1a/) |
|BA1B|rosalind.py|[Find the Most Frequent Words in a String](http://rosalind.info/problems/ba1b/) |
|BA1C|rosalind.py|[Find the Reverse Complement of a String](http://rosalind.info/problems/ba1c/) |
|BA1D|rosalind.py|[Find All Occurrences of a Pattern in a String](http://rosalind.info/problems/ba1d/) |
|BA1E|rosalind.py|[Find Patterns Forming Clumps in a String](http://rosalind.info/problems/ba1e/)|
|BA1F|rosalind.py|[Find a Position in a Genome Minimizing the Skew](http://rosalind.info/problems/ba1f/)|
|BA1G|rosalind.py|[Compute the Hamming Distance Between Two Strings](http://rosalind.info/problems/ba1g/)|
|BA1H|rosalind.py|[Find All Approximate Occurrences of a Pattern in a String](http://rosalind.info/problems/ba1h/) |
|BA1I|rosalind.py|[Find the Most Frequent Words with Mismatches in a String](http://rosalind.info/problems/ba1i/) |
|BA1J|rosalind.py|[Find Frequent Words with Mismatches and Reverse Complements](http://rosalind.info/problems/ba1j/) |
|BA1K|rosalind.py|[Generate the Frequency Array of a Strings](http://rosalind.info/problems/ba1f/) |
|BA1L|rosalind.py|[Implement PatternToNumber](http://rosalind.info/problems/ba1l/) |
|BA1M|rosalind.py|[Implement NumberToPattern](http://rosalind.info/problems/ba1m/) |
|BA1N|rosalind.py|[Generate the d-Neighborhood of a String](http://rosalind.info/problems/ba1n/) |
|2|-|How do we sequence Antibiotics?|
|BA2A|rosalind.py|[Implement MotifEnumeration](http://rosalind.info/problems/ba2a/) |
|BA2B |rosalind.py|[Find a Median String ](http://rosalind.info/problems/ba1j/)
|BA2C |rosalind.py|[Find a Profile-most Probable k-mer in a String	](http://rosalind.info/problems/ba2c/)
|BA2D |rosalind.py|[Implement GreedyMotifSearch](http://rosalind.info/problems/ba2d/)
|BA2E |rosalind.py|[Implement GreedyMotifSearch with Pseudocounts](http://rosalind.info/problems/ba2e/)
|BA2F |rosalind.py|[Implement RandomizedMotifSearch](http://rosalind.info/problems/ba2f/)	
|BA2G |rosalind.py|[Implement GibbsSampler](http://rosalind.info/problems/ba2g/)
|BA2H |rosalind.py|[Implement DistanceBetweenPatternAndStrings](http://rosalind.info/problems/ba2h/) 
|3|-|Which DNA Patterna play the Role of Molecular Clocks?|
|BA3A|rosalind.py|[Generate the k-mer Composition of a String](http://rosalind.info/problems/ba3a/) |
|BA3B|rosalind.py|[Reconstruct a String from its Genome Path](http://rosalind.info/problems/ba3a/)
|BA3C|rosalind.py|[Construct the Overlap Graph of a Collection of k-mers](http://rosalind.info/problems/ba3a/)
|BA3D |rosalind.py|[Construct the De Bruijn Graph of a String ](http://rosalind.info/problems/ba3a/)
|BA3E |rosalind.py|[Construct the De Bruijn Graph of a Collection of k-mers](http://rosalind.info/problems/ba3a/)
|BA3F |rosalind.py|[Find an Eulerian Cycle in a Graph](http://rosalind.info/problems/ba3a/)
|BA3G |rosalind.py|[Find an Eulerian Path in a Graph](http://rosalind.info/problems/ba3a/)
|BA3H |rosalind.py|[Reconstruct a String from its k-mer Composition](http://rosalind.info/problems/ba3a/)
|BA3I |rosalind.py|[Find a k-Universal Circular String](http://rosalind.info/problems/ba3a/)
|BA3J |rosalind.py|[Reconstruct a String from its Paired Composition](http://rosalind.info/problems/ba3a/)
|4|-|How do we assemble Genomes?|
|BA4A|rosalind.py|[Translate an RNA String into an Amino Acid String](http://rosalind.info/problems/ba4a/) |
|BA4B|rosalind.py|[Find Substrings of a Genome Encoding a Given Amino Acid String](http://rosalind.info/problems/ba4b/) |
|BA4C|rosalind.py|[Generate the Theoretical Spectrum of a Cyclic Peptide](http://rosalind.info/problems/ba4c/) |
|BA4D|rosalind.py|[Compute the Number of Peptides of Given Total Mass](http://rosalind.info/problems/ba4d/) |
|BA4E|rosalind.py|[Find a Cyclic Peptide with Theoretical Spectrum Matching an Ideal Spectrum](http://rosalind.info/problems/ba4e/) |
|BA4F|rosalind.py|[Compute the Score of a Cyclic Peptide Against a Spectrum](http://rosalind.info/problems/ba4f/) |
|BA4G|rosalind.py|[Implement LeaderboardCyclopeptideSequencing ](http://rosalind.info/problems/ba4g/) |
|BA4H|rosalind.py|[Generate the Convolution of a Spectrum ](http://rosalind.info/problems/ba4h/) |
|BA4I|rosalind.py|[Implement ConvolutionCyclopeptideSequencing](http://rosalind.info/problems/ba4i/) |
|BA4J|rosalind.py|[Generate the Theoretical Spectrum of a Linear Peptide](http://rosalind.info/problems/ba4j/) |
|BA4K|rosalind.py|[Compute the Score of a Linear Peptide ](http://rosalind.info/problems/ba4k/) |
|BA4L|rosalind.py|[Trim a Peptide Leaderboard](http://rosalind.info/problems/ba4l/) |
|BA4M|ba4m.py| [Solve the Turnpike Problem](http://rosalind.info/problems/ba4m/) |
|5|align.py|How do we compare Biological Sequences?|
|BA5A|align.py|[Find the Minimum Number of Coins Needed to Make Change](http://rosalind.info/problems/ba5a/)|
|BA5B|align.py|[Find the Length of a Longest Path in a Manhattan-like Grid](http://rosalind.info/problems/ba5b/)|
|BA5C|align.py|[Find a Longest Common Subsequence of Two Strings](http://rosalind.info/problems/ba5c/)|
|BA5D|align.py|[Find the Longest Path in a DAG](http://rosalind.info/problems/ba5d/)|
|BA5E|align.py|[Find a Highest-Scoring Alignment of Two Strings](http://rosalind.info/problems/ba5e/)|
|BA5F|align.py|[Find a Highest-Scoring Local Alignment of Two Strings](http://rosalind.info/problems/ba5f/)|
|BA5G|align.py| [Compute the Edit Distance Between Two Strings](http://rosalind.info/problems/ba5g/) |
|BA5H|BA5H.py| [Find a Highest-Scoring Fitting Alignment of Two Strings](http://rosalind.info/problems/ba5h/) |   
|BA5I|BA5I.py|[Find a Highest-Scoring Overlap Alignment of Two Strings](http://rosalind.info/problems/ba5i/)|    
|BA5J|BA5J.py|[Align Two Strings Using Affine Gap Penalties](http://rosalind.info/problems/ba5j/)|    
|BA5K|TBP|[Find a Middle Edge in an Alignment Graph in Linear Space](http://rosalind.info/problems/ba5k/) |
|BA5L|TBP|[Align Two Strings Using Linear Space](http://rosalind.info/problems/ba5l/) |    
|BA5M|TBP|[Find a Highest-Scoring Multiple Sequence Alignment](http://rosalind.info/problems/ba5m/) | 
|BA5N |align.py|[Find a Topological Ordering of a DAG](http://rosalind.info/problems/ba5n/) | 
|6|fragile.py|Are there fragile regions in the human genome?|
|BA6A|BA6A.py|[Implement GreedySorting to Sort a Permutation by Reversals ](http://rosalind.info/problems/ba6a/)|
|BA6B|BA6B.py|[Compute the Number of Breakpoints in a Permutation](http://rosalind.info/problems/ba6b/)|
|BA6C|BA6C.py|[Compute the 2-Break Distance Between a Pair of Genomes](http://rosalind.info/problems/ba6c/)|
|BA6D|TBP|[Find a Shortest Transformation of One Genome into Another by 2-Breaks](http://rosalind.info/problems/ba6d/)|
|BA6E|fragile.py| [Find All Shared k-mers of a Pair of Strings ](http://rosalind.info/problems/ba6e/)|
|BA6F|BA6F.py|[Implement Chromosome to Cycle](http://rosalind.info/problems/ba6f/)|
|BA6G|BA6G.py|[Implement Cycle to Chromosome](http://rosalind.info/problems/ba6g/)|
|BA6H|BA6H.py|[Implement Coloured Edges](http://rosalind.info/problems/ba6h/)|
|BA6I|BA6I.py|[Implement GraphToGenome ](http://rosalind.info/problems/ba6i/)|
|BA6J|BA6J.py|[Implement 2-BreakOnGenomeGraph ](http://rosalind.info/problems/ba6j/)|
|BA6K|BA6K| [Implement 2-BreakOnGenome ](http://rosalind.info/problems/ba6k/) (WIP)|
|7|-|Which animal gave us SARS?|
|BA7A|BA7A.py|[Compute Distances Between Leaves](http://rosalind.info/problems/ba7a/)|
|BA7B|BA7B.py|[Limb Length Problem](http://rosalind.info/problems/ba7b/)|
|BA7C|BA7C.py|[Implement Additive Phylogeny](http://rosalind.info/problems/ba7c/)|
|BA7D|BA7D.py|[Implement UPGMA](http://rosalind.info/problems/ba7d/)|
|BA7E|BA7E.py|[Implement the Neighbor Joining Algorithm](http://rosalind.info/problems/ba7e/)|
|BA7F|BA7F.py|[Implement SmallParsimony](http://rosalind.info/problems/ba7f/)|
|BA7G|BA7G.py|[Adapt SmallParsimony to Unrooted Trees](http://rosalind.info/problems/ba7g/)|
|8|-|How did Yeast become a Winemaker?|
|BA8A|BA8A.py | [Implement FarthestFirstTraversal](http://rosalind.info/problems/ba8a/) |
|BA8B|BA8B.py | [Compute the Squared Error Distortion](http://rosalind.info/problems/ba8b/) |
|BA8C|BA8C.py | [Implement the Lloyd Algorithm for k-Means Clustering](http://rosalind.info/problems/ba8c/) |
|BA8D|BA8D.py | [Implement the Soft k-Means Clustering Algorithm](http://rosalind.info/problems/ba8d/) |
|BA8E|BA8E.py | [Implement Hierarchical Clustering](http://rosalind.info/problems/ba8e/) |
|9|-|How do we locate disease causing mutations?|
|BA9A|BA9A.py | [Construct a Trie from a Collection of Patterns](http://rosalind.info/problems/ba9a/) |
|BA9B|BA9B.py | [Implement TrieMatching](http://rosalind.info/problems/ba9b/) |
|10|-|Why have biologists still not developed an HIV vaccine?|
|BA10A|BA10A.py hmm.py| [Compute the Probability of a Hidden Path](http://rosalind.info/problems/ba10a/) |
|BA10B|BA10B.py hmm.py| [Compute the Probability of an Outcome Given a Hidden Path](http://rosalind.info/problems/ba10b/) |
|BA10C|BA10C.py hmm.py| [Implement the Viterbi Algorithm](http://rosalind.info/problems/ba10c/) |
|11|-|Was T.rex just a big chicken?|
|BA11A|BA11A.py spectrum.py| [Spectrum Graph Construction](http://rosalind.info/problems/ba11a/) |
|BA11B|BA11B.py spectrum.py| [Implement DecodingIdealSpectrum](http://rosalind.info/problems/ba11b/) |
|BA11C|BA11C.py spectrum.py| [Convert a Peptide into a Peptide Vector ](http://rosalind.info/problems/ba11c/) |
|BA11D|BA11D.py spectrum.py| [Convert a Peptide Vector into a Peptide ](http://rosalind.info/problems/ba11d/) |

## Other Rosalind Problems

| Name | File|Description |
| ---- | ----------- | ------------------------------------------------|
|2sum|2sum.py|[2SUM](http://rosalind.info/problems/2sum/) |
|3sum|3sum.py|[3SUM](http://rosalind.info/problems/3sum/) |
|afrq|rosalind.py|[Counting Disease Carriers](http://rosalind.info/problems/afrq/) |
|asmq|ASMQ.py|[Assessing Assembly Quality with N50 and N75](http://rosalind.info/problems/asmq/) |
|aspc|rosalind.py|[Introduction to Alternative Splicing](http://rosalind.info/problems/aspc/) |
|bf|bf.py graphs.py|[Bellman-Ford Algorithm](http://rosalind.info/problems/bf/)|
|bfs|bfs.py|[Beadth First search](http://rosalind.info/problems/bfs/) |
|bins|BINS.py|[Binary Search](http://rosalind.info/problems/bfs/) |
|bip|bip.py graphs.py|[Testing Bipartiteness](http://rosalind.info/problems/bip/) |
|cat|rosalind.py|[Catalan Numbers and RNA Secondary Structures ](http://rosalind.info/problems/cat/) (WIP)|
|cc|cc.py graphs.py|[Connected Components](http://rosalind.info/problems/cc/) |
|cons|rosalind.py|[Consensus and Profile](http://rosalind.info/problems/cons/) |
|corr|CORR.py|[Error Correction in Reads](http://rosalind.info/problems/corr/) |
|cstr|cstr.py|[Creating a Character Table from Genetic Strings](http://rosalind.info/problems/cstr/)|
|ctbl|ctbl.py|[Creating a Character Table](http://rosalind.info/problems/ctbl/) |
|cte|cte.py graphs.py|[Shortest Cycle Through a Given Edge](http://rosalind.info/problems/cte/)|
|dag|dag.py graphs.py|[Testing Acyclicity](http://rosalind.info/problems/dag/) |
|dbru|DBRU.py|[Constructing a De Bruijn Graph](http://rosalind.info/problems/dbru/) |
|ddeg|DDEG.py graphs.py|[Double-Degree Array](http://rosalind.info/problems/ddeg/) |
|deg|DEG.py graphs.py|[Degree Array](http://rosalind.info/problems/deg/) |
|dfs| graphs.py|[Depth First Search](http://rosalind.info/problems/dfs/) |
|dij|dij.py graphs.py|[Dijkstra's Algorithm](http://rosalind.info/problems/dij/) |
|dna|rosalind.py|[Counting DNA Nucleotides](http://rosalind.info/problems/dna/) |
|ebin|rosalind.py|[Wright-Fisher's Expected Behavior](http://rosalind.info/problems/ebin/) |
|edit|edit.py|[Edit Distance](http://rosalind.info/problems/edit/) |
|edta|edta.py|[Edit Distance Alignment](http://rosalind.info/problems/edta/)|
|eval|rosalind.py|[Expected Number of Restriction Sites](http://rosalind.info/problems/eval/) |
|fibd|rosalind.py|[Mortal Fibonacci Rabbits](http://rosalind.info/problems/fibd/) |
|fib|rosalind.py|[Rabbits and Recurrence Relations](http://rosalind.info/problems/fib/) |
|foun|rosalind.py|[The Founder Effect and Genetic Drift](http://rosalind.info/problems/foun/) |
|frmt|FRMT.py|[Data Formats](http://rosalind.info/problems/frmt/)|
|gaff|gaff.py|[Global Alignment with Scoring Matrix and Affine Gap Penalty](http://rosalind.info/problems/gaff/) (WIP)|
|gasm|GASM.py|[Genome Assembly Using Reads](http://rosalind.info/problems/gasm/) |
|gbk|GBK.py|[GenBank Introduction](http://rosalind.info/problems/gbk/) |
|gcon|gcon.py|[Global Alignment with Constant Gap Penalty](http://rosalind.info/problems/gcon/)|
|gc|rosalind.py|[Computing GC Content](http://rosalind.info/problems/gc/) |
|glob|GLOB.py|[Global Alignment with Scoring Matrix](http://rosalind.info/problems/glob/) -- test harness only|
|grep|GREP.py|[Genome Assembly with Perfect Coverage and Repeats](http://rosalind.info/problems/grep/)|
|grph|rosalind.py|[Overlap Graphs](http://rosalind.info/problems/grph/) |
|gs|gs.py graphs.py|[General Sink](http://rosalind.info/problems/gs/)|
|hamm|rosalind.py|[Counting Point Mutations](http://rosalind.info/problems/hamm/) |
|hdag|hdag.py graphs.py|[Hamiltonian Path in DAG](http://rosalind.info/problems/hdag/) |
|iev|rosalind.py| [Calculating Expected Offspring](http://rosalind.info/problems/iev/)|
|indq|rosalind.py|[Independent Segregation of Chromosomes](http://rosalind.info/problems/indq/) |
|ins|INS.py|[Insertion Sort](http://rosalind.info/problems/ins/) |
|inv|INV.py|[Counting Inversions](http://rosalind.info/problems/inv/) |
|iprb|rosalind.py|[Mendel's First Law](http://rosalind.info/problems/iprb/) |
|kmer|rosalind.py|[Generalizing GC-Content](http://rosalind.info/problems/kmer/) |
|kmp|kmp.py|[Speeding Up Motif Finding](http://rosalind.info/problems/kmp/) |
|lcsm|rosalind.py|[Finding a Shared Motif](http://rosalind.info/problems/lcsm/) |
|lcsq|lcsq.py|[Finding a Shared Spliced Motif](http://rosalind.info/problems/lcsq/) |
|lgis|lgis.py|[Longest Increasing Subsequence](http://rosalind.info/problems/lgis/) |
|lia|rosalind.py|[Independent Alleles](http://rosalind.info/problems/lia/) |
|ling|ling.py|[Linguistic Complexity of a Genome](http://rosalind.info/problems/ling/) (WIP)|
|loca|loca.py|[Local Alignment with Scoring Matrix ](http://rosalind.info/problems/loca/) |
|long|LONG.py|[Genome Assembly as Shortest Superstring](http://rosalind.info/problems/long/) |
|med|MED.py|[Median](http://rosalind.info/problems/med/)|
|mend|mend.py|[Inferring a Genotype from a Pedigree](http://rosalind.info/problems/mend/) |
|mer|MER.py|[Merge Two Sorted Arrays](http://rosalind.info/problems/mer/) |
|mrna|rosalind.py|[Inferring mRNA from Protein](http://rosalind.info/problems/mrna/) |
|ms|MS.py|[Mergesort](http://rosalind.info/problems/ms/)|
|nwck|nwck.py|[Distances in Trees](http://rosalind.info/problems/nwck/) |
|nwc|nwc.py|[Negative Weight Cycle](http://rosalind.info/problems/nwc/) (wip)|
|oap|oap.py|[Overlap Alignment](http://rosalind.info/problems/oap/) (wip)|
|orfr|orfr.py|[Finding Genes with Open Reading Frames](http://rosalind.info/problems/orfr/) |
|orf|orf.py|[Open Reading Frames](http://rosalind.info/problems/orf/)|
|par3|PAR3.py|[3-Way Partition](http://rosalind.info/problems/par3/) |
|par|PAR.py|[2-Way Partition](http://rosalind.info/problems/par/) |
|pcov|PCOV.py|[Genome Assembly with Perfect Coverage](http://rosalind.info/problems/pcov/) |
|pdst|rosalind.py|[Creating a Distance Matrix ](http://rosalind.info/problems/pdst/) |
|pmch|pmch.py|[Perfect Matchings and RNA Secondary Structures](http://rosalind.info/problems/pmch/) |
|pper|rosalind.py|[Partial Permutations](http://rosalind.info/problems/pper/) |
|prot|rosalind.py|[Translating RNA into Protein](http://rosalind.info/problems/prot/) |
|ps|ps.py|[Partial sort](http://rosalind.info/problems/ps/)|
|ptra|PTRA.py|[Protein Translation](http://rosalind.info/problems/ptra/) |
|qs|QS.py|[Quicksort](http://rosalind.info/problems/qs/)|
|rear|rear.py|[Reversal distance](http://rosalind.info/problems/rear/) (WIP)|
|revc|rosalind.py| [Complementing a Strand of DNA](http://rosalind.info/problems/revc/)|
|revp|rosalind.py| [Locating Restriction Sites](http://rosalind.info/problems/revp/)|
|rna|rosalind.py|[Transcribing DNA into RNA](http://rosalind.info/problems/rna/) |
|rstr|rosalind.py|[Matching Random Motifs](http://rosalind.info/problems/rstr/) |
|scc|scc.py graphs.py|[Strongly Connected Components](http://rosalind.info/problems/scc/)|
|sc|sc.py graphs.py|[Semi-Connected Graph](http://rosalind.info/problems/sc/)|
|sexl|rosalind.py|[Sex-Linked Inheritance ](http://rosalind.info/problems/sexl/) |
|splc|rosalind.py|[RNA Splicing](http://rosalind.info/problems/splc/) |
|sq|sq.py graphics.py|[Square in a Graph](http://rosalind.info/problems/sq/)|
|sseq|rosalind.py|[Finding a Spliced Motif](http://rosalind.info/problems/sseq/) |
|subs|rosalind.py|[Finding a Motif in DNA](http://rosalind.info/problems/subs/) |
|tran|rosalind.py|[Transitions and Transversions](http://rosalind.info/problems/tran/) |
|trie|rosalind.py|[Introduction to Pattern Matching](http://rosalind.info/problems/trie/) |
|ts|ts.py align.py|[Topological sort](http://rosalind.info/problems/ts/)|
|wfmd|rosalind.py|[The Wright-Fisher Model of Genetic Drift](http://rosalind.info/problems/wfmd/) |
  
## Support

| Name |Description |
| -------------------------- | ------------------------------------------------|
|align.py|Common functions for [alignment](http://rosalind.info/problems/topics/alignment/) |
|graphs.py|Common functions for [graphs](http://rosalind.info/problems/topics/graphs/) |
|helpers.py|Utilities for formatting output, parsing input, etc |
|hmm.py|Hidden Markov Models|
|LICENSE|License Agreement|
|newick.py|Parser for files in [Newick format](http://evolution.genetics.washington.edu/phylip/newicktree.html) |
|README.md|This file|
|rosalind.py|Shared code|
|rosalind.wpr|WingWare Project File |
|spectrum.py|Utilites for mass spectroscopy|

## Other stuff

| Name | Description |
| ------------------------- | ------------------------------------------------|
| count-kmer-occurrences.py | Determine number of times each kmer appears in DNA, allowing for reverse complements and d-neighbourhoods |
