# Bioinformatics

Code inspired by [Finding Hidden Messages in DNA (Bioinformatics I)](https://class.coursera.org/hiddenmessages-003).

| Name | Description |
| -------------------------- | ------------------------------------------------| 
| ASMQ.py | 	Assessing Assembly Quality with N50 and N75 |
| BA8A.py | 	Implement FarthestFirstTraversal |
| BA8B.py | 	Compute the Squared Error Distortion |
| BA8C.py | Implement the Lloyd Algorithm for k-Means Clustering  |
| BA8D.py | 	Implement the Soft k-Means Clustering Algorithm |
| BINS.py | 	 	Binary Search |
| CORR.py | 	Error Correction in Reads  | 
| DBRU.py | 	 	Constructing a De Bruijn Graph |
| DDEG.py | 	 	Double-Degree Array |
| DEG.py | 	 	Degree Array |
| FRMT.py | 	Added copy |right notice 	4 days ago|
| GASM.py | 	 	Genome Assembly Using Reads |
| GBK.py | 	Added copy |right notice 	4 days ago|
| GREP.py | 	Solved GREP 	14 minutes ago|
| INS.py | 	 	Insertion Sort |
| INV.py | 	Added copy |right notice 	4 days ago|
| LICENSE |	License Agreement|
| LONG.py | 	Genome Assembly as Shortest Superstring |
| MED.py | 	Median|
| MER.py | 	Added copy |right notice 	4 days ago|
| MS.py | 	Added copy |right notice 	4 days ago|
| PAR.py | 	Added copy |right notice 	4 days ago|
| PAR3.py | 	Added copy |right notice 	4 days ago|
| PCOV.py | 	 Genome Assembly with Perfect Coverage |
| PTRA.py | 	Added copy |right notice 	4 days ago|
| QS.py | 	Quicksort|
| README.md 	Update README.md 	2 months ago|
| count-kmer-occurrences.py | Determine number of times each kmer appears in DNA, allowing for reverse complements and d-neighbourhoods |
| mend.py | 	 Inferring a Genotype from a Pedigree |
| newick.py | 	 | Parser for files in Newick format |
| nwck.py | 	Distances in Trees |
| rear.py | Reversal distance |
| rosalind.py | 	Shared code|
| rosalind.wpr | 	WingWare Project File |
