# Bioinformatics

Code inspired by [Finding Hidden Messages in DNA (Bioinformatics I)](https://class.coursera.org/hiddenmessages-003) and from [Rosalind](http://rosalind.info). 
NB: functions generally used zero based indexing; Rosalind uses 1-based.

| Name | Description |
| -------------------------- | ------------------------------------------------|
| 2sum.py | [2SUM](http://rosalind.info/problems/2sum/) |
| ASMQ.py | 	Assessing Assembly Quality with N50 and N75 |
| BA4M.py |   Solve the Turnpike Problem  |
| BA7A.py | Compute Distances Between Leaves |
| BA7B.py | Limb Length Problem |
| BA7C.py | Implement Additive Phylogeny |
| BA7D.py |  Implement UPGMA |
| BA7E.py | Implement the Neighbor Joining Algorithm |
| BA7F.py | Implement SmallParsimony |
| BA7G.py | Adapt SmallParsimony to Unrooted Trees |
| BA8A.py | 	Implement FarthestFirstTraversal |
| BA8B.py | 	Compute the Squared Error Distortion |
| BA8C.py | Implement the Lloyd Algorithm for k-Means Clustering  |
| BA8D.py | 	Implement the Soft k-Means Clustering Algorithm |
| BA8E.py | Implement Hierarchical Clustering |
| bfs.py  | 	Beadth First search |
| bip.py | [Testing Bipartiteness](http://rosalind.info/problems/bip/)|
| BINS.py | 	 	Binary Search |
| cc.py |	[Connected Components](http://rosalind.info/problems/cc/) |
| CORR.py | 	Error Correction in Reads  | 
| count-kmer-occurrences.py | Determine number of times each kmer appears in DNA, allowing for reverse complements and d-neighbourhoods |
| ctbl.py | Creating a Character Table |
| dag.py | [Testing Acyclicity](http://rosalind.info/problems/dag/) |
| DBRU.py | 	 	Constructing a De Bruijn Graph |
| DDEG.py | 	 	Double-Degree Array |
| DEG.py | 	 	Degree Array |
| FRMT.py | 	Data Formats|
| GASM.py | 	 	Genome Assembly Using Reads |
| GBK.py | 	GenBank Introduction |
| GREP.py | 	 Genome Assembly with Perfect Coverage and Repeats|
| helpers.py | Utilities for formatting output, parsing input, etc |
| INS.py | 	 	Insertion Sort |
| INV.py | 	Counting Inversions (WIP)|
| kmp.py | Speeding Up Motif Finding |
| LICENSE |	License Agreement|
| LONG.py | 	Genome Assembly as Shortest Superstring |
| maj.py |  [Majority Element](http://rosalind.info/problems/maj/) ||
| MED.py | 	Median|
| mend.py | 	 Inferring a Genotype from a Pedigree |
| MER.py |  	Merge Two Sorted Arrays |
| MS.py | 	Mergesort|
| newick.py | 	 | Parser for files in Newick format |
| nwck.py | 	Distances in Trees |
| PAR.py | 	 	2-Way Partition |
| PAR3.py | 	 	3-Way Partition |
| PCOV.py | 	 Genome Assembly with Perfect Coverage |
| pmch.py| 	Perfect Matchings and RNA Secondary Structures |
| PTRA.py | 	 	Protein Translation |
| QS.py | 	Quicksort|
| README.md 	|This file|
| rear.py | Reversal distance |
| rosalind.py | 	Shared code|
| rosalind.wpr | 	WingWare Project File |
