# Bioinformatics

Code inspired by [Finding Hidden Messages in DNA (Bioinformatics I)](https://class.coursera.org/hiddenmessages-003) and from [Rosalind](http://rosalind.info). 
NB: functions generally used zero based indexing; Rosalind uses 1-based.

## Textbook track

| # | Location | Description |
| ---- | -------------------------- | ------------------------------------------------|
| BA4M| ba4m.py |   [Solve the Turnpike Problem](http://rosalind.info/problems/ba4m/)  |
|BA5A|rosalind-old.py|[Compute the Edit Distance Between Two Strings](http://rosalind.info/problems/ba5a/)|
|BA5B|rosalind-old.py|[Compute the Edit Distance Between Two Strings](http://rosalind.info/problems/ba5b/)|
|BA5C|rosalind-old.py|[Compute the Edit Distance Between Two Strings](http://rosalind.info/problems/ba5c/)|
|BA5D|rosalind-old.py|[Compute the Edit Distance Between Two Strings](http://rosalind.info/problems/ba5d/)|
|BA5E|rosalind-old.py|[Compute the Edit Distance Between Two Strings](http://rosalind.info/problems/ba5e/)|
|BA5F|rosalind-old.py|[Compute the Edit Distance Between Two Strings](http://rosalind.info/problems/ba5f/)|
| BA5G | edit.py | [Compute the Edit Distance Between Two Strings](http://rosalind.info/problems/ba5g/) |
| BA7A | .py | [Compute Distances Between Leaves](http://rosalind.info/problems/ba7a/) |
| BA7B |.py | [Limb Length Problem](http://rosalind.info/problems/ba7b/) |
| BA7C| .py | [Implement Additive Phylogeny](http://rosalind.info/problems/ba7c/) |
| BA7D|.py |  [Implement UPGMA](http://rosalind.info/problems/ba7d/) |
| BA7E|.py | [Implement the Neighbor Joining Algorithm](http://rosalind.info/problems/ba7e/) |
| BA7F|.py | [Implement SmallParsimony](http://rosalind.info/problems/ba7f/) |
| BA7G|.py | [Adapt SmallParsimony to Unrooted Trees](http://rosalind.info/problems/ba7g/) |
| BA8A|.py | [Implement FarthestFirstTraversal](http://rosalind.info/problems/ba8a/) |
| BA8B|.py | [Compute the Squared Error Distortion](http://rosalind.info/problems/ba7b/) |
| BA8C|.py | [Implement the Lloyd Algorithm for k-Means Clustering](http://rosalind.info/problems/ba8c/)  |
| BA8D|.py | [Implement the Soft k-Means Clustering Algorithm](http://rosalind.info/problems/ba8d/) |
| BA8E|.py | [Implement Hierarchical Clustering](http://rosalind.info/problems/ba8e/) |

## Other problems



| Name | Description |
| -------------------------- | ------------------------------------------------|
| 2sum.py | [2SUM](http://rosalind.info/problems/2sum/) |
| ASMQ.py | 	[Assessing Assembly Quality with N50 and N75](http://rosalind.info/problems/asmq/) |
| bfs.py  | [Beadth First search](http://rosalind.info/problems/bfs/) |
| bip.py | [Testing Bipartiteness](http://rosalind.info/problems/bip/)|
| BINS.py | [Binary Search](http://rosalind.info/problems/bfs/) |
| cat.pt | [Catalan Numbers and RNA Secondary Structures ](http://rosalind.info/problems/cat/) (WIP)|
| cc.py |	[Connected Components](http://rosalind.info/problems/cc/) |
| CORR.py | [Error Correction in Reads](http://rosalind.info/problems/corr/)  | 
| count-kmer-occurrences.py | Determine number of times each kmer appears in DNA, allowing for reverse complements and d-neighbourhoods |
| ctbl.py | [Creating a Character Table](http://rosalind.info/problems/ctbl/)  |
| dag.py | [Testing Acyclicity](http://rosalind.info/problems/dag/) |
| DBRU.py |	[Constructing a De Bruijn Graph](http://rosalind.info/problems/ctbl/) |
| DDEG.py | 	[Double-Degree Array](http://rosalind.info/problems/ddeg/) |
| DEG.py | 	 	[Degree Array](http://rosalind.info/problems/deg/) |
|edit.py |	[Edit Distance](http://rosalind.info/problems/edit/) |
| edta.py | [Edit Distance Alignment](http://rosalind.info/problems/edta/) (WIP)|
| FRMT.py | 	[Data Formats](http://rosalind.info/problems/frmt/)|
| GASM.py | 	 	[Genome Assembly Using Reads](http://rosalind.info/problems/gasm/) |
| GBK.py | 	[GenBank Introduction](http://rosalind.info/problems/gbk/) |
| GREP.py | 	 [Genome Assembly with Perfect Coverage and Repeats](http://rosalind.info/problems/grep/)|
| helpers.py | Utilities for formatting output, parsing input, etc |
| INS.py | 	 	[Insertion Sort](http://rosalind.info/problems/ins/) |
| INV.py | 	[Counting Inversions](http://rosalind.info/problems/inv/) (WIP)|
| kmp.py | [Speeding Up Motif Finding](http://rosalind.info/problems/kmp/) |
| lcsq.py | 	[Finding a Shared Spliced Motif](http://rosalind.info/problems/lcsq/) |
| LICENSE |	License Agreement|
| LONG.py | 	[Genome Assembly as Shortest Superstring](http://rosalind.info/problems/long/) |
| maj.py |  [Majority Element](http://rosalind.info/problems/maj/) ||
| MED.py | 	[Median](http://rosalind.info/problems/med/)|
| mend.py | 	 [Inferring a Genotype from a Pedigree](http://rosalind.info/problems/mend/) |
| MER.py |  	[Merge Two Sorted Arrays](http://rosalind.info/problems/mer/) |
| MS.py | 	[Mergesort](http://rosalind.info/problems/ms/)|
| newick.py 	 | Parser for files in Newick format |
| nwck.py | 	[Distances in Trees](http://rosalind.info/problems/nwck/) |
| PAR.py | 	 [2-Way Partition](http://rosalind.info/problems/par/) |
| PAR3.py | 	 [3-Way Partition](http://rosalind.info/problems/par3/) |
| PCOV.py | 	 [Genome Assembly with Perfect Coverage](http://rosalind.info/problems/pcov/) |
| pmch.py| 	[Perfect Matchings and RNA Secondary Structures](http://rosalind.info/problems/pmch/) |
| PTRA.py | 	 [Protein Translation](http://rosalind.info/problems/ptra/) |
| QS.py | 	[Quicksort](http://rosalind.info/problems/qs/)|
| README.md 	|This file|
| rear.py | [Reversal distance](http://rosalind.info/problems/rear/) |
| rosalind.py | 	Shared code|
| rosalind.wpr | 	WingWare Project File |
